snippets/lazy_segment_tree.cpp

```diff
@@ -0,0 +1,126 @@
+const long long INF = 1e18 + 9;
+
+template<typename segment, typename update>
+class segmenttree{
+public:
+  struct node {
+    segment seg;
+    update upd;
+  };
+
+  int sz;
+  vector<node> tree;
+
+  segmenttree(){}
+  segmenttree(int n){
+    init(n);
+  }
+  void init(int n){
+    sz = 1;
+    while(sz < n){
+      sz <<= 1;
+    }
+    tree.assign(2 * sz, node());
+  }
```

Comment:
can we have the constructor to build the segtree on an array? Or it doesn't matter?

Reply:
just insert elements using update operation.. works in O(nlogn) which should not be a problem !!

Reply:
sure.. thanks